KAFKA-14959: remove delayed queue and exempt sensors during ClientQuota and ClientRequestQuota managers shutdown

machi1990 opened the pull request:
Follows up on https://github.com/apache/kafka/pull/13623#discussion_r1182592921

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

machi1990 committed:
KAFKA-14959: remove delayed queue and exempt sensors during ClientQuota and ClientRequestQuota managers shutdown

machi1990 commented:
@rajinisivaram @dajac please review when you've the chance

machi1990 commented:
I am wondering if these sensors that are dynamically created should be removed as well: https://github.com/apache/kafka/blob/ca11a87e86e8e3c65043f747f35cae770b1efb7c/core/src/main/scala/kafka/server/ClientQuotaManager.scala#L393-L402 looking for suggestion here

machi1990 commented:
@divijvaidya since you worked on https://github.com/apache/kafka/pull/13623, would you be open to give this PR a review? Thanks

divijvaidya commented on core/src/main/scala/kafka/server/ClientRequestQuotaManager.scala:
```diff
@@ -28,13 +28,12 @@ import org.apache.kafka.server.quota.ClientQuotaCallback
   val DefaultInactiveExemptSensorExpirationTimeSeconds = Long.MaxValue
 
-  private val ExemptSensorName = "exempt-" + QuotaType.Request
+  val ExemptSensorName = "exempt-" + QuotaType.Request
```
this could be `private[server]`

divijvaidya commented on core/src/test/scala/unit/kafka/server/ClientQuotaManagerTest.scala:
```diff
@@ -418,6 +418,17 @@ class ClientQuotaManagerTest extends BaseClientQuotaManagerTest {
   }
 
+  @Test
+  def testDelayedQueueSensorShouldShouldExistAfterInstantiationAndBeRemovedAfterShutdown(): Unit = {
```
typo

\shouldshould\should

machi1990 replied:
good catch, I'll update.

divijvaidya commented on core/src/test/scala/unit/kafka/server/ClientQuotaManagerTest.scala:
```diff
@@ -418,6 +418,17 @@ class ClientQuotaManagerTest extends BaseClientQuotaManagerTest {
+    val clientQuotaManager = new ClientQuotaManager(config, metrics, Produce, time, "")
+    var delayedQueueSensor = metrics.getSensor(sensorName)
+    assertNotNull(delayedQueueSensor, "delayed queue sensor should exist")
+    clientQuotaManager.shutdown()
```
please ensure that this gets cleaned up at the end of the test if the above assertion fails.

(same comment for other test)

machi1990 replied:
Good call.

divijvaidya commented on core/src/test/scala/unit/kafka/server/ClientQuotaManagerTest.scala:
```diff
@@ -418,6 +418,17 @@ class ClientQuotaManagerTest extends BaseClientQuotaManagerTest {
+    var delayedQueueSensor = metrics.getSensor(sensorName)
+    assertNotNull(delayedQueueSensor, "delayed queue sensor should exist")
+    clientQuotaManager.shutdown()
+    delayedQueueSensor = metrics.getSensor(sensorName)
```
Another validation we can perform is to ensure that
`assertTrue(metrics.metrics().isEmpty)` at `@AfterEach` of BaseClientQuotaManagerTest

machi1990 replied:
I looked onto this and apparently the metrics.metrics() List is never empty because this metric https://github.com/apache/kafka/blob/ca11a87e86e8e3c65043f747f35cae770b1efb7c/clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java#L180 isn't at the moment removed when we `Metrics#close()` is called.

Is that to be expected? If not, I can create a separate JIRA and address it separate from this PR.

divijvaidya commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -572,6 +572,7 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
   }
 
   def shutdown(): Unit = {
+    metrics.removeSensor(delayQueueSensor.name())
```
we should clean up the metric at the very end of the shutdown call (inside a finally). This is because the metric may still get values while the thread hasn't completed shutdown.

machi1990 replied:
I'll adjust. Thanks

divijvaidya reviewed (changes requested):
Thank you for the PR. It's worth noting that the ClientQuotaManagers are shutdown on broker shutdown and hence, it's not a big deal if we are leaking metrics right now. Nevertheless, it's good practice to close metrics properly and remove them from the JMX server.

machi1990 committed:
KAFKA-14959: Address review comments

machi1990 commented:
> Thank you for the PR. It's worth noting that the ClientQuotaManagers are shutdown on broker shutdown and hence, it's not a big deal if we are leaking metrics right now. Nevertheless, it's good practice to close metrics properly and remove them from the JMX server.

Thanks for the review @divijvaidya I've addressed all of your comments except for one where I've left a suggested course of action. Let me know what you think

divijvaidya commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -574,6 +574,7 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
   def shutdown(): Unit = {
     initiateShutdown()
     throttledChannelReaper.awaitShutdown()
+    metrics.removeSensor(delayQueueSensor.name())
```
in a finally block please.

machi1990 replied:
You are right!

divijvaidya commented on core/src/main/scala/kafka/server/ClientRequestQuotaManager.scala:
```diff
@@ -92,4 +91,9 @@ class ClientRequestQuotaManager(private val config: ClientQuotaManagerConfig,
+
+  override def shutdown(): Unit = {
+    super.shutdown()
+    metrics.removeSensor(exemptSensor.name())
```
same. in a finally block please.

divijvaidya commented on core/src/main/scala/kafka/server/ClientRequestQuotaManager.scala:
```diff
@@ -28,13 +28,12 @@ import org.apache.kafka.server.quota.ClientQuotaCallback
 import scala.jdk.CollectionConverters._
 
 object ClientRequestQuotaManager {
-  val QuotaRequestPercentDefault = Int.MaxValue.toDouble
```
please investigate why do we have this unused variable. Asking because it could be a bug where we want to use this but are not using this,

machi1990 replied:
The const was defining the default value for quota request percent: Before #10025 this was done in the DynamicConfig.scala class.

After #10025, we stopped using by just hardcoding the value: https://github.com/mumrah/kafka/blob/a9ede6eaedc1cdef2757bd2f03de1cb0efe170f8/clients/src/main/java/org/apache/kafka/common/config/internals/QuotaConfigs.java#L62

divijvaidya replied:
ack. thank you.

divijvaidya commented:
I wanted to let you know that this is on my radar and I need a few more days to get to this.

machi1990 commented:
> I wanted to let you know that this is on my radar and I need a few more days to get to this.

Thanks @divijvaidya

machi1990 commented:
I've limited bandwidth thus I won't be able to carry this work forward. I'll happily close this MR and unassign myself from the JIRA for someone else to take it over.
